core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileOutputWriter.scala

```diff
@@ -216,3 +217,44 @@ object ProfileOutputWriter {
     sb.toString()
   }
 }
+
+/**
+ * For a given App ID, this class prevents older attempt IDs from writing to the output if a newer
+ * attempt ID is found.
+ *
+ * Throws OlderAttemptException if app is older attempt.
+ */
+class AttemptAwareProfileOutputWriter(app: AppBase, outputDir: String,
+      filePrefix: String, numOutputRows: Int, outputCSV: Boolean = false)
+  extends ProfileOutputWriter(outputDir, filePrefix, numOutputRows, outputCSV) {
+
+  override def writeText(strToWrite: String): Unit = {
+    if (app.isOlderAttempt) {
+      throw OlderAttemptException(app)
+    }
+    super.writeText(strToWrite)
+  }
+
+  override def writeSparkRapidsBuildInfo(headerText: String,
+      sparkRapidsBuildInfo: Seq[SparkRapidsBuildInfoEvent]): Unit = {
+    if (app.isOlderAttempt) {
+      throw OlderAttemptException(app)
+    }
+    super.writeSparkRapidsBuildInfo(headerText, sparkRapidsBuildInfo)
+  }
+
+  override def write(headerText: String, outRows: Seq[ProfileResult],
+      emptyTableText: Option[String] = None, tableDesc: Option[String] = None): Unit = {
+    if (app.isOlderAttempt) {
+      throw OlderAttemptException(app)
+    }
+    super.write(headerText, outRows, emptyTableText, tableDesc)
+  }
+}
+
+object AttemptAwareProfileOutputWriter {
+  def from(app: AppBase, pWriter: ProfileOutputWriter): AttemptAwareProfileOutputWriter = {
+    new AttemptAwareProfileOutputWriter(app, pWriter.outputDir, pWriter.filePrefix,
+      pWriter.numOutputRows, pWriter.outputCSV)
+  }
+}
```

Comment:
I don't quite understand this object. For example, a thread can be writing to a file. A new ttempt passes all those checks and start writing at the same time. We cannot be sure what would be final file content in such scenario.
What we need to do is avoid two threads writing to the directory at the same time. This can be done using synchronization/mutexes. Either an active thread is interrupted or the newer attempt waits for the older thread to finish writing.